### PR TITLE
Fix #117 and add related test cases

### LINKED ISSUE #117: dpkt http wrong parse_headers logic
_From [almog1...@gmail.com](https://code.google.com/u/106388495642043329325/) on May 21, 2012 05:52:27_

I have been encountered a problem which I found also some people complaining about on the net.

The description on the error from here http://stackoverflow.com/questions/6337878/parsing-pcap-files-with-dpkt-python is about the dpkt.http.parse_headers function
"""
While it seems to effectively parse most of the packets, I'm receiving a NeedData("premature end of headers") exception on some. They appear to be valid packets within WireShark, so I'm a bit confused as to why the exceptions are being thrown.
"""

I investigated the problem and found that the exception is raised if the HTTP is not ending with "\r\n\r\n".

I have researched more than 10 Milion Http Request and there wasn't even one packet where this exception was relevant.

I changed the parse headers on my installatioin to look like this

def parse_headers(f):
    """Return dict of HTTP headers parsed from a file object."""
    d = {}
    while 1:
        line = f.readline()
        if not line:
            break
        \# I commented  this line out because the way this parser works is wrong.
        \# When the http does not finish with '\r\n\r\n' the parsing is faild!
        \#    raise dpkt.NeedData('premature end of headers')

        \# The next two liness are responsibble for stopping the parsing when the
        \# line equals '\r\n'
        line = line.strip()
        if not line:
            break
        l = line.split(None, 1)
        if not l[0].endswith(':'):
            raise dpkt.UnpackError('invalid header: &#37;r' &#37; line)
        k = l[0][:-1].lower()
        v = len(l) != 1 and l[1] or ''
        if k in d:
            if not type(d[k]) is list:
                d[k] = [d[k]]
            d[k].append(v)
        else:
            d[k] = v
    return d

and its working great.

Many thanks,
Almog

_Original issue: http://code.google.com/p/dpkt/issues/detail?id=90_

### PR DESCRIPTION
**Pull request by sunhao2014**
Fix #117, add comments to the key logic and add test cases provided by @brifordwylie.

### CODE REVIEW

File: dpkt/http.py [hunk 1]
```diff
@@ -247,6 +252,40 @@ def test_request_version():
     except:
         pass
 
+def test_invalid_header():
+    # valid header.
+    s = 'POST /main/redirect/ab/1,295,,00.html HTTP/1.0\r\n' \
```
Comment (kbandla) [inline dpkt/http.py]: This is okay too, but you can also write long strings like:
s = ( 'aaaaaa'
        'bbbbbbb'
)

**Comment by brifordwylie** [comment]
look good merging. Also lets all start using https://waffle.io/kbandla/dpkt for issue management.
1) Looking/thinking about an issue? Put it 'On Deck'
2) Working on an issue? Put it 'In Progress' and make sure you assign yourself.

The boxes can just be dragged to the different columns... it also updates/syncs with GitHub so it works pretty nice.

File: dpkt/http.py (hunk 1 above)
Comment (sunhao2014) [inline dpkt/http.py]: Yeah, hopefully for this case we'll get an exception. I'll add this case to the unit test.

**Comment by sunhao2014** [comment]
Cool. I tried the waffle. It's really a great issue tracking tool.

File: dpkt/http.py (hunk 1 above)
Comment (sunhao2014) [inline dpkt/http.py]: @kbandla, done! Please see 36b81cd.

**Pull request merged**